File: native/src/edf/gel_pl.cpp
#include "tasks.h"

#include "edf/gel_pl.h"

#include <vector>
#include <limits>
#include <algorithm>
#include <cmath>
#include <iostream>

static bool reversed_order(const fractional_t& first,
                           const fractional_t& second) {
    return second < first;
}

GELPl::GELPl(unsigned int num_processors, const TaskSet& ts,
             unsigned int num_rounds)
:no_cpus(num_processors), tasks(ts), rounds(num_rounds)
{
    fractional_t sys_utilization;
    tasks.get_utilization(sys_utilization);
    // Compute ceiling
    integral_t util_ceil_pre = sys_utilization.get_num();
    mpz_cdiv_q(util_ceil_pre.get_mpz_t(),
               sys_utilization.get_num().get_mpz_t(),
               sys_utilization.get_den().get_mpz_t());
    util_ceil = util_ceil_pre.get_ui();
    std::vector<unsigned long> prio_pts;
    fractional_t S = 0;
    std::vector<fractional_t> Y_ints;

    int task_count = tasks.get_task_count();
    // Reserve capacity in all vectors to minimize allocation costs.
    prio_pts.reserve(task_count);
    Y_ints.reserve(task_count);
    S_i.reserve(task_count);
    G_i.reserve(task_count);

    // For faster lookups
    utilizations.reserve(task_count);
    for (int i = 0; i < task_count; i++) {
        utilizations.push_back(tasks[i].get_wcet());
        utilizations[i] /= tasks[i].get_period();
    }
    
    unsigned long min_prio_pt = std::numeric_limits<unsigned long>::max();

    // Compute initial priority points, including minimum.
    for (int i = 0; i < task_count; i++) {
        const Task& task = tasks[i];
        unsigned long new_prio_pt = task.get_prio_pt();
        prio_pts.push_back(new_prio_pt);
        if (new_prio_pt < min_prio_pt) {
            min_prio_pt = new_prio_pt;
        }
    }

    // Reduce to compute minimum.  Also compute Y intercepts, S_i values, and
    // S.
    for (int i = 0; i < task_count; i++) {
        prio_pts[i] -= min_prio_pt;
        const Task& task = tasks[i];
        unsigned long wcet = task.get_wcet();
        unsigned long period = task.get_period();
        S_i.push_back(prio_pts[i]);
        fractional_t& S_i_i = S_i[i];
        S_i_i *= -1;
        S_i_i /= period;
        S_i_i += 1;
        S_i_i *= wcet;
        if (S_i_i < 0) {
            S_i_i = 0;
        }
        S += S_i_i;
        Y_ints.push_back(wcet);
        fractional_t& Y_ints_i = Y_ints[i];
	Y_ints_i *= -1;
        Y_ints_i /= no_cpus;
        Y_ints_i *= utilizations[i];
        Y_ints_i += wcet;
        Y_ints_i -= S_i_i;
    }

    fractional_t s;
    if (rounds == 0) {
        compute_exact_s(S, Y_ints, s);
    }
    else {
        compute_binsearch_s(S, Y_ints, s);
    }

    for (int i = 0; i < task_count; i++) {
        fractional_t x_i = s;
        fractional_t x_comp = tasks[i].get_wcet();
        x_comp /= no_cpus;
        x_i -= x_comp;
        // Compute ceiling
        integral_t xi_ceil = x_i.get_num();
        mpz_cdiv_q(xi_ceil.get_mpz_t(),
                   x_i.get_num().get_mpz_t(),
                   x_i.get_den().get_mpz_t());
        bounds.push_back(prio_pts[i]
                         + tasks[i].get_wcet()
                         + xi_ceil.get_ui());
        G_i.push_back(s);
        G_i[i] *= utilizations[i];
        G_i[i] += Y_ints[i];
    }
}

void GELPl::compute_exact_s(const fractional_t& S,
                            const std::vector<fractional_t>& Y_ints,
                            fractional_t& s) {
    int task_count = tasks.get_task_count();

    std::vector<ReplacementType> replacements;
    for (int i = 0; i < task_count; i++) {
        for (int j = i + 1; j < task_count; j++) {
            // We can ignore parallel and identical lines - either don't
            // intersect or we don't care which is picked.
            if (utilizations[i] != utilizations[j]) {
                fractional_t intersect_den = utilizations[i];
                intersect_den -= utilizations[j];
                fractional_t intersect = Y_ints[j];
                intersect -= Y_ints[i];
                intersect /= intersect_den;
                ReplacementType replacement;
                replacement.location = intersect;
                if (intersect >= 0) {
                    if (utilizations[i] < utilizations[j]) {
                        replacement.old_task = i;
                        replacement.old_task_utilization = utilizations[i];
                        replacement.new_task = j;
                    }
                    else {
                        replacement.old_task = j;
                        replacement.old_task_utilization = utilizations[j];
                        replacement.new_task = i;
                    }
                    replacements.push_back(replacement);
                }
            }
        }
    }
    std::sort(replacements.begin(), replacements.end());

    std::vector<bool> task_pres;
    task_pres.assign(task_count, false);
    
    fractional_t current_value = S;
    fractional_t current_slope = no_cpus;
    current_slope *= -1;

    std::vector<TaggedValue> init_pairs;
    init_pairs.reserve(task_count);
    for (int i = 0; i < task_count; i++) {
    	TaggedValue new_pair;
        new_pair.task = i;
        new_pair.value = Y_ints[i];
        init_pairs.push_back(new_pair);
    }
    
    // Only if we have tasks contributing to G
    if (util_ceil >= 2) {
        // Allows us to efficiently compute sum of top m-1 elements.  They may
        // not be in order but must be the correct choices.
        std::nth_element(init_pairs.begin(),
                         init_pairs.begin() + util_ceil - 2,
                         init_pairs.end());

        for (int i = 0; i < util_ceil - 1; i++) {
            unsigned int task_index = init_pairs[i].task;
            task_pres[task_index] = true;
            current_value += init_pairs[i].value;
            current_slope += utilizations[task_index];
        }
    }
    
    unsigned int rindex = 0;
    fractional_t next_s = 0;
    s = 1;
    while (s > next_s) {
        fractional_t current_s = next_s;
        s = current_value;
        s /= current_slope;
        s *= -1;
        s += current_s;
        if (rindex < replacements.size()) {
            ReplacementType replacement = replacements[rindex];
            next_s = replacement.location;
            fractional_t val_inc = next_s;
            val_inc -= current_s;
            val_inc *= current_slope;
            current_value += val_inc;
            // Apply replacement, if appropriate
            if (task_pres[replacement.old_task]
                    && !task_pres[replacement.new_task]) {
                task_pres[replacement.old_task] = false;
                current_slope -= utilizations[replacement.old_task];
                task_pres[replacement.new_task] = true;
                current_slope += utilizations[replacement.new_task];
            }
            rindex++;
        }
        else {
            next_s = s;
            next_s += 1;
        }
    }
    // At this point, "s" should be the appropriate return value
}

void GELPl::compute_binsearch_s(const fractional_t& S,
                                const std::vector<fractional_t>& Y_ints,
				fractional_t& s) {
    fractional_t min_s = 0;
    fractional_t max_s = 1;
    while (!M_lt_0(max_s, S, Y_ints)) {
        min_s = max_s;
        max_s *= 2;
    }
    
    for (int i = 0; i < rounds; i++) {
        fractional_t middle = min_s;
	middle += max_s;
	middle /= 2;
        if (M_lt_0(middle, S, Y_ints)) {
            max_s = middle;
        }
        else {
            min_s = middle;
        }
    }

    // max_s is guaranteed to be a legal bound.
    s = max_s;
}

bool GELPl::M_lt_0(const fractional_t& s, const fractional_t& S,
                   const std::vector<fractional_t>& Y_ints) {
    std::vector<fractional_t> Gvals;
    int task_count = tasks.get_task_count();
    for (int i = 0; i < task_count; i++) {
        Gvals.push_back(utilizations[i]);
        Gvals[i] *= s;
        Gvals[i] += Y_ints[i];
    }

    fractional_t final_val = no_cpus;
    final_val *= -1;
    final_val *= s;
    final_val += S;
    
    // Only if there will be tasks contributing to G
    if (util_ceil >= 2) {
        // Again, more efficient computation by not totally sorting.
        std::nth_element(Gvals.begin(),
                         Gvals.begin() + util_ceil - 2,
                         Gvals.end(),
                         reversed_order);
        for (int i = 0; i < util_ceil - 1; i++) {
            final_val += Gvals[i];
        }
    }

    return (final_val < 0);
}

File: schedcat/sched/edf/gel_pl.py
"""This module computes tardiness bounds for G-EDF-like schedulers.  Currently
   G-EDF and G-FL (See Erickson and Anderson ECRTS'12) are supported.  This
   module works by analyzing compliant vectors (see Erickson, Devi, Baruah
   ECRTS'10), analyzing the involved piecewise linear functions (see Erickson,
   Guan, Baruah OPODIS'10).  Notation is used as in EA'12.

   The analysis is general enough to support other GEL schedulers trivially.
"""

from __future__ import division

from math import ceil

from fractions import Fraction

from schedcat.util.quantor import forall

import schedcat.sched
if schedcat.sched.using_native:
    import schedcat.sched.native as native

import heapq

class AnalysisDetails:
    def __init__(self, tasks, gel_obj = None):
        num_tasks = len(tasks)
        self.tasks = tasks
        if gel_obj is not None:
            self.bounds = [gel_obj.get_bound(i) for i in range(num_tasks)]
            self.S_i = [gel_obj.get_Si(i) for i in range(num_tasks)]
            self.G_i = [gel_obj.get_Gi(i) for i in range(num_tasks)]

    def max_lateness(self):
        return max([self.bounds[i] - self.tasks[i].deadline
                    for i in range(len(self.tasks))])

def compute_gfl_response_details(no_cpus, tasks, rounds):
    """This function computes response time bounds for the given set of tasks
       using G-FL.  "rounds" determines the number of rounds of binary search;
       using "0" results in using the exact algorithm.
    """
    for task in tasks:
        task.pp = task.deadline - int((no_cpus - 1) / (no_cpus) * task.cost)
    return compute_response_details(no_cpus, tasks, rounds)

def compute_gedf_response_details(no_cpus, tasks, rounds):
    """This function computes response time bounds for a given set of tasks
       using G-EDF.  "rounds" determines the number of rounds of binary search;
       using "0" results in using the exact algorithm.
    """
    for task in tasks:
        task.pp = task.deadline
    return compute_response_details(no_cpus, tasks, rounds)

def compute_response_details(no_cpus, tasks, rounds):
    if (no_cpus == 1) and forall(tasks)(lambda t: t.pp == t.period):
        details = AnalysisDetails(tasks)
        details.bounds = [task.period for task in tasks]
        details.S_i = [0.0 for task in tasks]
        details.G_i = [0.0 for task in tasks]
        return details
    if schedcat.sched.using_native:
        native_ts = schedcat.sched.get_native_taskset(tasks)
        gel_obj = native.GELPl(no_cpus, native_ts, rounds)
        return AnalysisDetails(tasks, gel_obj)
    else:
        return compute_response_bounds(no_cpus, tasks, rounds)

def compute_response_bounds(no_cpus, tasks, rounds):
    """This function computes response time bounds for the given set of tasks
       and priority points.  "rounds" determines the number of rounds of binary
       search; using "0" results in using the exact algorithm.
    """

    if not has_bounded_tardiness(no_cpus, tasks):
        return None

    # Compute utilization ceiling
    util_ceil = int(ceil(tasks.utilization_q()))
    # First uniformly reduce scheduler priority points to derive analysis
    # priority points.  Due to uniform reduction, does not change scheduling
    # decisions.  Shown in EA'12 to improve bounds.
    sched_pps = [task.pp for task in tasks]
    min_priority_point = min(sched_pps)
    analysis_pps = [sched_pps[i] - min_priority_point
                    for i in range(len(sched_pps))]

    #Initialize S_i and Y_ints
    S_i = [Fraction(0)]*len(tasks)
    Y_ints = [Fraction(0)]*len(tasks)
    utilizations = [Fraction(task.cost, task.period) for task in tasks];
    # Calculate S_i and y-intercept (for G(\vec{x}) contributors)
    for i, task in enumerate(tasks):
        S_i[i] = max(Fraction(0), Fraction(task.cost) * (Fraction(1)
                     - Fraction(analysis_pps[i], task.period)))
        Y_ints[i] = (Fraction(0) - Fraction(task.cost, no_cpus)) * \
                     utilizations[i] + Fraction(task.cost) - S_i[i]

    if rounds == 0:
        s = compute_exact_s(no_cpus, tasks, util_ceil, sum(S_i), Y_ints,
                            utilizations)
    else:
        s = compute_binsearch_s(no_cpus, tasks, util_ceil, sum(S_i), Y_ints,
                                utilizations, rounds)

    details = AnalysisDetails(tasks)
    details.bounds = [int(ceil(s - Fraction(tasks[i].cost, no_cpus) + 
                      Fraction(tasks[i].cost) + Fraction(analysis_pps[i])))
                      for i in range(len(tasks))]
    details.S_i = S_i
    details.G_i = [Y_ints[i] + s * utilizations[i]
                   for i in range(len(tasks))]
    return details

def compute_exact_s(no_cpus, tasks, util_ceil, S, Y_ints, utilizations):
    replacements = []
    for i, task1 in enumerate(tasks):
        for j in range(i+1, len(tasks)):
            task2 = tasks[j]
            # Parallel lines do not intersect, and choice of identical lines
            # doesn't matter.  Ignore all such cases.
            if utilizations[i] != utilizations[j]:
                intersect = Fraction(Y_ints[j] - Y_ints[i],
                                     utilizations[i] - utilizations[j])
                if intersect >= 0:
                    if utilizations[i] < utilizations[j]:
                        replacements.append( (intersect, i, j) )
                    else:
                        replacements.append( (intersect, j, i) )
    # Break ties by order of increasing utilization of replaced tasks.  Avoids
    # an edge case.  Consider tasks A, B, C, in order of decreasing
    # utilization.  If the top m-1 items include tasks B and C, it is possible
    # (without the tie break) to have the following replacements:
    # C->B (no change)
    # B->A (now A and C in set considered)
    # C->A (no change)
    #
    # However, proper behavior should include A and B in considered set.  The
    # tie break avoids this case.
    replacements.sort(key=lambda r: (r[0], utilizations[r[1]]))

    # The piecewise linear function we are tracing is G(s) + S(\tau) - ms, as
    # discussed (with L(s) in place of G(s)) in EGB'10.  We start at s = 0 and
    # trace its value each time a slope changes, hoping for a value of zero.

    # While tracing piecewise linear function, keep track of whether each task
    # contributes to G(\vec{x}).  Array implementation allows O(1) lookups and
    # changes.
    task_pres = [False]*len(tasks)
    
    # Initial value and slope.
    current_value = S
    current_slope = Fraction(-1 * no_cpus)

    init_pairs = heapq.nlargest(util_ceil - 1, enumerate(Y_ints),
                                key=lambda p: p[1])

    # For s = 0, just use Y-intercepts to determine what is present.
    for pair in init_pairs:
        task_pres[pair[0]] = True
        current_value += pair[1]
        current_slope += utilizations[pair[0]]

    # Index of the next replacement
    rindex = 0
    next_s = Fraction(0)
    zero = next_s + Fraction(1)
    while zero > next_s:
        current_s = next_s
        zero = current_s - Fraction(current_value, current_slope)
        if rindex < len(replacements):
            replacement = replacements[rindex]
            next_s = replacement[0]
            current_value += (next_s - current_s) * current_slope
            # Apply replacement, if appropriate.
            if task_pres[replacement[1]] and not task_pres[replacement[2]]:
                task_pres[replacement[1]] = False
                current_slope -= utilizations[replacement[1]]
                task_pres[replacement[2]] = True
                current_slope += utilizations[replacement[2]]
            rindex += 1
        else:
            next_s = zero + 1
    return zero

def compute_binsearch_s(no_cpus, tasks, util_ceil, S, Y_ints, utilizations,
                        rounds):
    def M(s):
        Gvals = heapq.nlargest(util_ceil - 1,
                               [Y_ints[i] + utilizations[i] * s
                                for i in range(len(tasks))])
        return sum(Gvals) + S - Fraction(no_cpus) * s

    min_s = Fraction(0)
    max_s = Fraction(1)
    while M(max_s) > 0:
        min_s = max_s
        max_s *= Fraction(2)

    for i in range(rounds):
        middle = Fraction(max_s + min_s, 2)
        if M(middle) < 0:
            max_s = middle
        else:
            min_s = middle

    #max_s is guaranteed to be a legal bound.
    return max_s

def has_bounded_tardiness(no_cpus, tasks):
    return tasks.utilization() <= no_cpus and \
        forall(tasks)(lambda t: t.period >= t.cost)

def bound_gedf_response_times(no_cpus, tasks, rounds):
    response_details = compute_gedf_response_details(no_cpus, tasks, rounds)

    return bound_response_times(tasks, response_details)

def bound_gfl_response_times(no_cpus, tasks, rounds):
    response_details = compute_gfl_response_details(no_cpus, tasks, rounds)

    return bound_response_times(tasks, response_details)

def bound_response_times(tasks, response_details):
    if response_details is None:
        return False

    else:
        for i in range(len(tasks)):
            tasks[i].response_time = response_details.bounds[i]
        return True

